src/types.rs

```diff
@@ -281,165 +205,153 @@ impl BoundedStorable for PrincipalStorable {
-    #[serde(rename = "cyclesPerMessageByte")]
-    pub cycles_per_message_byte: u64,
-}
-
-#[derive(Clone, CandidType, Deserialize)]
-pub struct UpdateProviderArgs {
-    #[serde(rename = "providerId")]
-    pub provider_id: u64,
-    #[serde(rename = "credentialPath")]
-    pub credential_path: Option<String>,
-    #[serde(rename = "credentialHeaders")]
-    pub credential_headers: Option<Vec<HttpHeader>>,
-    #[serde(rename = "cyclesPerCall")]
-    pub cycles_per_call: Option<u64>,
-    #[serde(rename = "cyclesPerMessageByte")]
-    pub cycles_per_message_byte: Option<u64>,
-}
-
-#[derive(Clone, Debug, CandidType, Deserialize)]
-pub struct ManageProviderArgs {
-    #[serde(rename = "providerId")]
-    pub provider_id: u64,
-    #[serde(rename = "chainId")]
-    pub chain_id: Option<u64>,
-    pub primary: Option<bool>,
-    pub service: Option<RpcService>,
-}
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct ApiKey(String);
 
-#[derive(Clone, CandidType, Deserialize)]
-pub struct Provider {
-    #[serde(rename = "providerId")]
-    pub provider_id: u64,
-    pub owner: Principal,
-    #[serde(rename = "chainId")]
-    pub chain_id: u64,
-    pub hostname: String,
-    #[serde(rename = "credentialPath")]
-    pub credential_path: String,
-    #[serde(rename = "credentialHeaders")]
-    pub credential_headers: Vec<HttpHeader>,
-    #[serde(rename = "cyclesPerCall")]
-    pub cycles_per_call: u64,
-    #[serde(rename = "cyclesPerMessageByte")]
-    pub cycles_per_message_byte: u64,
-    #[serde(rename = "cyclesOwed")]
-    pub cycles_owed: u128,
-    pub primary: bool,
+impl ApiKey {
+    /// Explicitly read API key (use sparingly)
+    pub fn read(&self) -> &str {
+        &self.0
+    }
 }
 
-impl Provider {
-    pub fn api(&self) -> RpcApi {
-        RpcApi {
-            url: format!("https://{}{}", self.hostname, self.credential_path),
-            headers: if self.credential_headers.is_empty() {
-                None
-            } else {
-                Some(self.credential_headers.clone())
-            },
-        }
+// Enable printing data structures which include an API key
+impl fmt::Debug for ApiKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{API_KEY_REPLACE_STRING}")
     }
 }
 
-impl Storable for Metadata {
-    fn to_bytes(&self) -> Cow<[u8]> {
-        Cow::Owned(Encode!(self).unwrap())
-    }
-    fn from_bytes(bytes: Cow<[u8]>) -> Self {
-        Decode!(&bytes, Self).unwrap()
+impl TryFrom<String> for ApiKey {
+    type Error = String;
+    fn try_from(key: String) -> Result<ApiKey, Self::Error> {
+        validate_api_key(&key)?;
+        Ok(ApiKey(key))
     }
 }
 
-impl Storable for Provider {
+impl Storable for ApiKey {
     fn to_bytes(&self) -> Cow<[u8]> {
-        Cow::Owned(Encode!(self).unwrap())
+        self.0.to_bytes()
     }
+
     fn from_bytes(bytes: Cow<[u8]>) -> Self {
-        Decode!(&bytes, Self).unwrap()
+        Self(String::from_bytes(bytes))
```

Comment:
A concern with doing this is that if we change the validation logic to exclude already registered API keys, it could potentially break the canister on upgrade and require a fresh reinstall in another NNS proposal. We could probably accept that risk, but it seems like a tradeoff. WDYT?